File: 15-3sum/3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> ans;
        sort(nums.begin(), nums.end());
        // fix 1 and find the two with two pinter technique
        for (int i = 0; i < nums.size(); i++) {
            if (i > 0 && nums[i] == nums[i - 1])
                continue;
            int left = i + 1;
            int right = nums.size() - 1;
            while (left < right) {
                if (nums[left] + nums[right] + nums[i] > 0) {
                    right--;
                } else if (nums[left] + nums[right] + nums[i] < 0) {
                    left++;
                } else if (nums[left] + nums[right] + nums[i] == 0) {
                    ans.push_back({nums[left], nums[right], nums[i]});
                    // also check the next is a duplicate or not both in left
                    // and right and skip
                    while (left < right && nums[left] == nums[left + 1])
                        left++;
                    while (left < right && nums[right] == nums[right - 1])
                        right--;
                    left++;
                    right--;
                }
            }
        }
        return ans;
    }
};

File: 15-3sum/3sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function (nums) {
    let target = 0;
    // as we know we can do this in O(N^2) how we can fix one do the two pointer technique here
    let len = nums.length;
    // sort the nums as well
    let ans = [];
    nums.sort((a, b) => a - b);
    for (let i = 0; i < len - 2; i++) {
        // check for duplicates and skip
        if (i > 0 && nums[i] == nums[i - 1]) continue;
        let l = i + 1;
        let r = len - 1;
        while (l < r) {
            let sum = nums[l] + nums[r] + nums[i];
            if (sum == target) {
                ans.push([nums[l], nums[r], nums[i]])
                l++
                r--;
                while (nums[l] == nums[l - 1]) l++;
                while (nums[r] == nums[r + 1]) r--;
            }
            else if (sum < target) {
                l++
            }
            else if (sum > target) {
                r--
            }
        }
    }
    return ans;
};